### DISCUSSION TITLE
[Thread] What's wrong with wiring?

### DISCUSSION THREAD

**Discussion by RadCatish** [FakeFishGames/Barotrauma#4823] [Ideas and suggestions]
What makes wiring/circuitry frustrating/overly difficult, both in game and in editor?
There are so many posts here offering suggested fixes which often turn out to not be implementable. Perhaps if we collate our grievances here, someone with knowledge of the codebase could take a look and have an epiphany about how to change it? 😅 I'm sure many of our problems stem from similar features.

Feel free to ignore if this is a bad idea!

**Comment by JamesTCrenshaw** [comment]
I feel that the wiring problem in-editor stems from not really being an editor thing... it's the same as in-game, and is very time consuming.
in-game, wiring is often hidden and obfuscated. This makes it very difficult to fix any damage that has been done to the wiring system through grief or accident.

Another issue is the fact that a lot of in-game wiring jobs are profoundly ugly, with wires going across the ship in every which way.

Yet another issue is the 'engineer' in-game who takes it upon themself to rewire the entire ship with new systems, lights, airlocks, etc. This makes it very hard to 'reverse' to the original state, and forces you to accept the changes, or move to a different ship/campaign. (Due to the fact you don't know what the original state of the wiring was, and may be totally unable to fix what has been done.)

Most problems with the wiring don't actually stem from the wiring system, but how people use or interact with it. The wiring system is actually fairly good, although could use some better UI and QOL improvements from the editor.

**Comment by 3e849f2e5c** [comment]
As the "sub editor guy" who worked on 95% of the new submarine editor features after Silky Smooth I thought I'd give some insight. In short the current wiring system is a raging spaghetti monster that I've yet to tame.

I implemented the global undo/redo functionality in the halloween update and had to completely leave out wiring mode or the feature wouldn't have been released until 2023.

I do have a personal list of things I'd like to add to the wiring mode and I've added a lot of entries to it from this discussion board but I feel like every time I take a look at one of them I instantly get caught trying to untangle the wiring behavior and end up giving up.
Something as simple as making the wires always show on top in wiring mode took me couple hours of digging knee deep into the code and I contributed my own share of spaghetti to it by implementing rewiring (double clicking to pick a wire back up in sub editor) since I had to work around the inventory trying to destroy the wire nodes and the fact that wires are "one way" requiring me to reverse the nodes if trying to pick up the other end of the wire.
I'd love to one day just take the sub editor side of the wiring apart and rewrite it but I've got my hands full currently and there are other things to take care of with higher priority.

Also fun fact there is actually a invisible character in the sub editor that is constantly teleported under your cursor (the screwdriver you see on your cursor is actually the invisible character holding it) because the current wiring system could not function without a character and inventory.

**Reply by JamesTCrenshaw** [reply]
I think you've done a really good job. The wiring system is fairly robust, and although it may currently be a rat's nest of code that only works one way (the way it is) and isn't modular yet; it **_works._** And that is one hell of an accomplishment, considering all of the things that the wiring system can do _right now._ 

Galldren is one hell of a workaround to such a complex problem, too. Good job on thinking that one.

**Reply by someone972** [reply]
The more I work with the code, the more respect I have for the amount of work you guys are able to complete given some of the limitations of design in areas. Understanding wiring has been an area I've been looking to get into, so if there's something on your hit-list you don't have time to get around to, and you'd like some outside help, with feel free to hit me up on discord. Working in the codebase is a fun hobby, so if I can help out with anything I'm more than happy.

**Comment by Working-Joe** [comment]
Here's my opinion on the wiring system, based on a lot of experience in building circuits reading through all the (signal component) related code.

In my opinion the wire handling/circuit building of baro is pretty acceptable. The main gripe I have with the wiring system is that there are a lot of hidden behaviours/inconsistencies in how components are built. To name a few:

- The update order of components causes delays in signal propagation, which can lead to wrong values or incorrect signal flow. (See also the readme [on this repo](https://github.com/Jlobblet/Barotrauma-Circuit-Resolver))
- Some components output constantly on their update, others output as soon as they receive a signal.
- Time frames have different meanings on different components. Some components send their false output once they time out, others send nothing.
- Any arithmetic component (and RegEx components) for some reason (bug?) starts with their inputs true for the duration of their timeframe. This can not be patched easily because it currently provides the functionality of detecting a round start. 
- Some components (which send on their update) can receive multiple inputs in a single frame, but only process the last received ones before their update. 

It has proven rather difficult especially for newer players to even discover that signals are not constant and that components don't always send a signal. This leads to confusion especially with mechanics like toggle vs. set inputs. Time frames are also notoriously difficult to figure out without help. 

I am working on thinking up an elegant solution solving all of these problems which I hope to at some point present to the devs (or even PR, depending on how much time I can find), but there's a few details I'm still working out. 

(To be specific, I've got a good concept solution that would work intuitively where all components would behave the same while all the functionality would be very close to the current build. However, this concept would require all components to send out a signal on every frame which I think would be bad performance wise. Although this is already the case for the most used components in the current system, I'm hoping to figure out a way to make all components update only on signal edges.)

**Reply by Working-Joe** [reply]
Also as a sidenote I would still love to see a circuit-box like representation to make circuits look better in game and in general make circuits more manageable (see also #4721).

**Comment by T-Hutch** [comment]
I'll start by saying that I'm enjoying the hell out of learning electrical engineer.  If I had a single item on my 'want' list it would simply be the ability to move a wire from one destination to another without disconnecting both ends, retrieving the wire, and starting over.  This is especially hellish when it comes to wiring items on the outside of the ship or worse, hidden by hull.  If I move a connector from the screw/interface so that it's on the bottom of the panel, moving it to my inventory or double clicking should exit the panel and I'd be back in 'wiring mode'.  --  In a separate vein, I don't feel like having two grids coming from a reactor with the same overall load as a single grid should cause a failure in the grid with the low load.  It makes it very difficult to have battery power support different objects than the reactor and that's a little unrealistic.  --  Lastly, it seems odd that I can't wire several supercapacitors in serial so that they share load.

**Comment by someone972** [comment]
My biggest gripe has typically been with power delivery and the bugs that come with it. Junction boxes and relays don't always play nice together, and chains of things typically break even if they make sense "electrically". Also, Junction boxes don't sit well with my mental model atm. Since everything connects at one point, how does it even know what the load is? Talking in real-world terms, that's no different than not having the junction box at all and just twisting all the wires together. Seems like it would make more sense if it had a power-in and power out leads.
